File: src/main/java/org/team1/nbe1_3_team01/global/auth/jwt/service/JwtService.java
package org.team1.nbe1_3_team01.global.auth.jwt.service;

import io.jsonwebtoken.Claims;
import io.jsonwebtoken.ExpiredJwtException;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.SignatureAlgorithm;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.team1.nbe1_3_team01.global.auth.jwt.respository.RefreshTokenRepository;
import org.team1.nbe1_3_team01.global.auth.jwt.token.RefreshToken;

import java.util.Date;
import java.util.Optional;

@Service
@RequiredArgsConstructor
@Slf4j
public class JwtService {

    @Value("${jwt.secretKey}")
    private String secretKey;

    @Value("${jwt.access.expiration}")
    private Long accessTokenExpirationPeriod;

    @Value("${jwt.refresh.expiration}")
    private Long refreshTokenExpirationPeriod;

    @Value("${jwt.access.header}")
    private String accessHeader;

    @Value("${jwt.refresh.header}")
    private String refreshHeader;

    private static final String ACCESS_TOKEN_SUBJECT = "AccessToken";
    private static final String REFRESH_TOKEN_SUBJECT = "RefreshToken";
    private static final String USERNAME_CLAIM = "username";
    private static final String BEARER = "Bearer ";

    private final RefreshTokenRepository refreshTokenRepository;

    /**
     * AccessToken 생성 메서드
     */
    public String createAccessToken(String username) {
        Date now = new Date();
        Claims claims = Jwts.claims();
        claims.put(USERNAME_CLAIM, username);
        claims.setExpiration(new Date(now.getTime() +accessTokenExpirationPeriod));
        return Jwts.builder()
                .setSubject(ACCESS_TOKEN_SUBJECT)
                .setClaims(claims)
                .signWith(SignatureAlgorithm.HS256, secretKey)
                .compact();
    }

    /**
     * RefreshToken 생성 메서드
     * 생성 후 refreshTokenRepository에 저장
     */

    public String createRefreshToken(String username) {
        Date now = new Date();
        Claims claims = Jwts.claims();
        claims.setExpiration(new Date(now.getTime() + refreshTokenExpirationPeriod));
        String token = Jwts.builder()
                .setSubject(REFRESH_TOKEN_SUBJECT)
                .setClaims(claims)
                .signWith(SignatureAlgorithm.HS256, secretKey)
                .compact();

        RefreshToken refreshToken = RefreshToken.builder()
                .username(username)
                .token(token)
                .build();

        refreshTokenRepository.save(refreshToken);
        return token;
    }

    /**
     * RefreshToken과 AccessToken을 응답 헤더에 넣어주는 메서드
     */
    public void sendAccessAndRefreshToken(HttpServletResponse response, String accessToken, String refreshToken) {
        response.setStatus(HttpServletResponse.SC_OK);
        setAccessTokenHeader(response, accessToken);
        setRefreshTokenHeader(response, refreshToken);
        log.info("Access Token, Refresh Token 헤더 설정 완료");
    }

    /**
     * Bearer 를 제거하여 refreshToken만 추출해 내는 메서드
     */

    public Optional<String> extractRefreshToken(HttpServletRequest request) {
        return Optional.ofNullable(request.getHeader(refreshHeader))
                .filter(refreshToken -> refreshToken.startsWith(BEARER))
                .map(refreshToken -> refreshToken.replace(BEARER, ""));
    }

    /**
     * Bearer 를 제거하여 accessToken만 추출해 내는 메서드
     */
    public Optional<String> extractAccessToken(HttpServletRequest request) {
        return Optional.ofNullable(request.getHeader(accessHeader))
                .filter(accessToken -> accessToken.startsWith(BEARER))
                .map(accessToken -> accessToken.replace(BEARER, ""));
    }

    /**
     * accessToken에서 username을 추출해 내는 메서드
     */
    public Optional<String> extractUsername(String accessToken) {
        try {
            Claims claims = Jwts.parser()
                    .setSigningKey(secretKey)
                    .parseClaimsJws(accessToken)
                    .getBody();

            return Optional.ofNullable(claims.get(USERNAME_CLAIM, String.class)); // username 클레임 가져오기
        } catch (Exception e) {
            log.error("액세스 토큰이 유효하지 않습니다.");
            return Optional.empty();
        }
    }

    /**
     * AccessToken을 응답 헤더에 넣어주는 메서드
     */
    private void setAccessTokenHeader(HttpServletResponse response, String accessToken) {
        response.setHeader(accessHeader, accessToken);
    }

    /**
     * RefreshToken을 응답 헤더에 넣어주는 메서드
     */
    private void setRefreshTokenHeader(HttpServletResponse response, String refreshToken) {
        response.setHeader(refreshHeader, refreshToken);
    }

    /**
     * 해당 토큰이 유효한지 검증하는 메서드
     */

    public void validateToken(String token) {
        try {
            Jwts.parserBuilder()
                    .setSigningKey(secretKey)
                    .build()
                    .parseClaimsJws(token);
        } catch (ExpiredJwtException e) {
            log.error("토큰이 만료되었습니다. {}", e.getMessage());
            throw e;
        }
        catch (Exception e) {
            log.error("유효하지 않은 토큰입니다. {}", e.getMessage());
            throw e;
        }
    }

}

File: src/main/java/org/team1/nbe1_3_team01/global/auth/jwt/service/JwtService.kt
package org.team1.nbe1_3_team01.global.auth.jwt.service

import io.jsonwebtoken.Jwts
import io.jsonwebtoken.SignatureAlgorithm
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import org.team1.nbe1_3_team01.global.auth.jwt.respository.RefreshTokenRepository
import org.team1.nbe1_3_team01.global.auth.jwt.token.RefreshToken
import java.security.Key
import java.util.*
import javax.crypto.spec.SecretKeySpec

@Service
class JwtService(
    private val refreshTokenRepository: RefreshTokenRepository
) {
    @Value("\${jwt.secretKey}")
    private lateinit var secretKey: String

    @Value("\${jwt.access.expiration}")
    private var accessTokenExpirationPeriod: Long? = null

    @Value("\${jwt.refresh.expiration}")
    private var refreshTokenExpirationPeriod: Long? = null

    @Value("\${jwt.access.header}")
    private lateinit var accessHeader: String

    @Value("\${jwt.refresh.header}")
    private lateinit var refreshHeader: String

    companion object {
        private const val ACCESS_TOKEN_SUBJECT = "AccessToken"
        private const val REFRESH_TOKEN_SUBJECT = "RefreshToken"
        private const val USERNAME_CLAIM = "username"
        private const val BEARER = "Bearer "
    }

    private fun getSigningKey(): Key {
        val decodedKey = Base64.getDecoder().decode(secretKey)
        return SecretKeySpec(decodedKey, SignatureAlgorithm.HS256.jcaName)
    }


    /**
     * AccessToken 생성 메서드
     */
    fun createAccessToken(username: String): String {
        val now = Date()
        val claims = Jwts.claims()
        claims[USERNAME_CLAIM] = username
        claims.setExpiration(Date(now.time + accessTokenExpirationPeriod!!))
        return Jwts.builder()
            .setSubject(ACCESS_TOKEN_SUBJECT)
            .setClaims(claims)
            .signWith(getSigningKey(), SignatureAlgorithm.HS256)
            .compact()
    }

    /**
     * RefreshToken 생성 메서드
     * 생성 후 refreshTokenRepository에 저장
     */
    fun createRefreshToken(username: String): String {
        val now = Date()
        val claims = Jwts.claims()
        claims.setExpiration(Date(now.time + refreshTokenExpirationPeriod!!))
        val token = Jwts.builder()
            .setSubject(REFRESH_TOKEN_SUBJECT)
            .setClaims(claims)
            .signWith(getSigningKey(), SignatureAlgorithm.HS256)
            .compact()

        val refreshToken = RefreshToken.of(
            username = username,
            token = token
        )

        refreshTokenRepository.save(refreshToken)
        return token
    }

    /**
     * RefreshToken과 AccessToken을 응답 헤더에 넣어주는 메서드
     */
    fun sendAccessAndRefreshToken(response: HttpServletResponse, accessToken: String, refreshToken: String) {
        response.status = HttpServletResponse.SC_OK
        setAccessTokenHeader(response, accessToken)
        setRefreshTokenHeader(response, refreshToken)
    }

    /**
     * Bearer 를 제거하여 refreshToken만 추출해 내는 메서드
     */
    fun extractRefreshToken(request: HttpServletRequest): String? {
        val refreshToken = request.getHeader(refreshHeader)
        return if (refreshToken != null && refreshToken.startsWith(BEARER)) {
            refreshToken.replace(BEARER, "")
        } else null

    }

    /**
     * Bearer 를 제거하여 accessToken만 추출해 내는 메서드
     */
    fun extractAccessToken(request: HttpServletRequest): String? {
        val accessToken = request.getHeader(accessHeader)
        return if (accessToken != null && accessToken.startsWith(BEARER)) {
            accessToken.replace(BEARER, "")
        } else null
    }

    /**
     * accessToken에서 username을 추출해 내는 메서드
     */
    fun extractUsername(accessToken: String?): String? {
        try {
            val claims = Jwts.parserBuilder()
                .setSigningKey(getSigningKey())
                .build()
                .parseClaimsJws(accessToken)
                .body

            return claims[USERNAME_CLAIM, String::class.java]
        } catch (e: Exception) {
            return null
        }
    }

    /**
     * AccessToken을 응답 헤더에 넣어주는 메서드
     */
    private fun setAccessTokenHeader(response: HttpServletResponse, accessToken: String) {
        response.setHeader(accessHeader, accessToken)
    }

    /**
     * RefreshToken을 응답 헤더에 넣어주는 메서드
     */
    private fun setRefreshTokenHeader(response: HttpServletResponse, refreshToken: String) {
        response.setHeader(refreshHeader, refreshToken)
    }

    /**
     * 해당 토큰이 유효한지 검증하는 메서드
     */
    fun validateToken(token: String) {
        try {
            Jwts.parserBuilder()
                .setSigningKey(getSigningKey())
                .build()
                .parseClaimsJws(token)
        } catch (e: Exception) {
            throw e
        }
    }

}
